File: internal/collector/system.go
package collector

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/jefferymvp/openwrt_status/internal/model"
)

// GetSystemInfo 获取系统基础状态概览
func GetSystemInfo() model.SystemInfo {
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "OpenWrt"
	}

	uptimeSec, uptimeStr := parseUptime()
	load1, load5, load15 := parseLoadAvg()
	totalMem, freeMem, availMem, usagePct := parseMemInfo()

	return model.SystemInfo{
		Hostname:       hostname,
		CurrentTime:    time.Now(),
		UptimeSeconds:  uptimeSec,
		UptimeFormat:   uptimeStr,
		LoadAvg1:       load1,
		LoadAvg5:       load5,
		LoadAvg15:      load15,
		TotalMemoryKB:  totalMem,
		FreeMemoryKB:   freeMem,
		AvailMemoryKB:  availMem,
		MemoryUsagePct: usagePct,
	}
}

// parseUptime 解析 /proc/uptime
func parseUptime() (int64, string) {
	content, err := os.ReadFile("/proc/uptime")
	if err != nil {
		return 0, "unknown"
	}

	fields := strings.Fields(string(content))
	if len(fields) == 0 {
		return 0, "unknown"
	}

	uptimeFloat, err := strconv.ParseFloat(fields[0], 64)
	if err != nil {
		return 0, "unknown"
	}

	totalSec := int64(uptimeFloat)
	days := totalSec / 86400
	hours := (totalSec % 86400) / 3600
	minutes := (totalSec % 3600) / 60
	seconds := totalSec % 60

	var format string
	if days > 0 {
		format = fmt.Sprintf("%d天 %d小时 %d分 %d秒", days, hours, minutes, seconds)
	} else if hours > 0 {
		format = fmt.Sprintf("%d小时 %d分 %d秒", hours, minutes, seconds)
	} else {
		format = fmt.Sprintf("%d分 %d秒", minutes, seconds)
	}

	return totalSec, format
}

// parseLoadAvg 解析 /proc/loadavg
func parseLoadAvg() (float64, float64, float64) {
	content, err := os.ReadFile("/proc/loadavg")
	if err != nil {
		return 0, 0, 0
	}

	fields := strings.Fields(string(content))
	if len(fields) < 3 {
		return 0, 0, 0
	}

	l1, _ := strconv.ParseFloat(fields[0], 64)
	l5, _ := strconv.ParseFloat(fields[1], 64)
	l15, _ := strconv.ParseFloat(fields[2], 64)

	return roundFloat(l1, 2), roundFloat(l5, 2), roundFloat(l15, 2)
}

// parseMemInfo 解析 /proc/meminfo
func parseMemInfo() (total uint64, free uint64, avail uint64, usagePct float64) {
	file, err := os.Open("/proc/meminfo")
	if err != nil {
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		valFields := strings.Fields(parts[1])
		if len(valFields) == 0 {
			continue
		}

		val, err := strconv.ParseUint(valFields[0], 10, 64)
		if err != nil {
			continue
		}

		switch key {
		case "MemTotal":
			total = val
		case "MemFree":
			free = val
		case "MemAvailable":
			avail = val
		}
	}

	if total > 0 {
		var used uint64
		if avail > 0 {
			used = total - avail
		} else {
			used = total - free
		}
		usagePct = roundFloat((float64(used)/float64(total))*100.0, 1)
	}

	return
}

File: src/collector/system.rs
use std::fs;

use chrono::Utc;

use super::round_float;
use crate::model::SystemInfo;

/// 获取系统基础状态概览
pub fn get_system_info() -> SystemInfo {
    let hostname = read_hostname();
    let (uptime_seconds, uptime_format) = parse_uptime();
    let (load_avg_1, load_avg_5, load_avg_15) = parse_loadavg();
    let (total_memory_kb, free_memory_kb, avail_memory_kb, memory_usage_percent) = parse_meminfo();

    SystemInfo {
        hostname,
        current_time: Utc::now(),
        uptime_seconds,
        uptime_format,
        load_avg_1,
        load_avg_5,
        load_avg_15,
        total_memory_kb,
        free_memory_kb,
        avail_memory_kb,
        memory_usage_percent,
    }
}

fn read_hostname() -> String {
    if let Ok(name) = fs::read_to_string("/proc/sys/kernel/hostname") {
        let trimmed = name.trim();
        if !trimmed.is_empty() {
            return trimmed.to_string();
        }
    }
    std::env::var("HOSTNAME")
        .or_else(|_| std::env::var("COMPUTERNAME"))
        .unwrap_or_else(|_| "OpenWrt".to_string())
}

fn parse_uptime() -> (i64, String) {
    let content = match fs::read_to_string("/proc/uptime") {
        Ok(c) => c,
        Err(_) => return (0, "unknown".to_string()),
    };

    let first_field = content.split_whitespace().next().unwrap_or("0");
    let total_sec = first_field.parse::<f64>().map(|f| f as i64).unwrap_or(0);

    let days = total_sec / 86400;
    let hours = (total_sec % 86400) / 3600;
    let minutes = (total_sec % 3600) / 60;
    let seconds = total_sec % 60;

    let format = if days > 0 {
        format!("{}天 {}小时 {}分 {}秒", days, hours, minutes, seconds)
    } else if hours > 0 {
        format!("{}小时 {}分 {}秒", hours, minutes, seconds)
    } else {
        format!("{}分 {}秒", minutes, seconds)
    };

    (total_sec, format)
}

fn parse_loadavg() -> (f64, f64, f64) {
    let content = match fs::read_to_string("/proc/loadavg") {
        Ok(c) => c,
        Err(_) => return (0.0, 0.0, 0.0),
    };

    let fields: Vec<&str> = content.split_whitespace().collect();
    if fields.len() < 3 {
        return (0.0, 0.0, 0.0);
    }

    let l1 = fields[0].parse::<f64>().unwrap_or(0.0);
    let l5 = fields[1].parse::<f64>().unwrap_or(0.0);
    let l15 = fields[2].parse::<f64>().unwrap_or(0.0);

    (round_float(l1, 2), round_float(l5, 2), round_float(l15, 2))
}

fn parse_meminfo() -> (u64, u64, u64, f64) {
    let content = match fs::read_to_string("/proc/meminfo") {
        Ok(c) => c,
        Err(_) => return (0, 0, 0, 0.0),
    };

    let mut total = 0u64;
    let mut free = 0u64;
    let mut avail = 0u64;

    for line in content.lines() {
        if let Some((key, val)) = line.split_once(':') {
            let key = key.trim();
            let num = val
                .split_whitespace()
                .next()
                .and_then(|s| s.parse::<u64>().ok())
                .unwrap_or(0);

            match key {
                "MemTotal" => total = num,
                "MemFree" => free = num,
                "MemAvailable" => avail = num,
                _ => {}
            }
        }
    }

    let mut usage_percent = 0.0;
    if total > 0 {
        let used = if avail > 0 {
            total.saturating_sub(avail)
        } else {
            total.saturating_sub(free)
        };
        usage_percent = round_float((used as f64 / total as f64) * 100.0, 1);
    }

    (total, free, avail, usage_percent)
}
